### PR TITLE
Fix String#scrub!

### PR DESCRIPTION
**Pull request by LillianZ**
Sometimes the TR `string#scrub! ` returns nil when the MRI implementation and the TR `string#scrub`  return a scrubbed string. This PR  fixes the behaviour of `string#scrub! ` without  impacting the behaviour of `string#scrub` (on the test suite that surfaced the issue with `string#scrub! `); and now calls `scrub!` in `scrub`, which seems more common than the other way around,  though I'm not entirely why that ordering would cause this issue.

### CODE REVIEW

**Comment by eregon** [comment]
Could you add a spec showing the problematic behavior?

From what I see neither `scrub` nor `scrub!` would ever return `nil` on TruffleRuby, and they don't seem to return `nil` on CRuby either (and in the docs).

File: src/main/ruby/truffleruby/core/string.rb [hunk 1]
```diff
@@ -984,7 +984,11 @@ def match?(pattern, pos=0)
 
   def scrub(replace = nil, &block)
     return dup if valid_encoding?
+    dup.scrub!(replace, &block)
+  end
 
+  def scrub!(replace = nil, &block)
+    return self if valid_encoding?
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/string.rb:991]: This seems a nice optimization for `scrub!` :+1:

**Comment by eregon** [comment]
Regarding which methods calls the other one, the current order (scrub! calling scrub) seems a little bit better because `"a\x81".scrub` would allocate a single String, vs 2 with this PR (dup + `Primitive.string_scrub`). Not very important though.
OTOH, the early check `return self if valid_encoding?` in this PR would work for both methods (with scrub calling scrub!), while `return dup if valid_encoding?` in `scrub` would still create an extra String for `"a".scrub!` (but we can have both early checks).

In general, most String operations can't be done in place, even more so since we use Ropes, so it tends to be more natural to have primitives returning a new String than mutating the current String inplace.

I think the main reason the pattern `foo = dup.foo!` is used is that some `!` methods return `nil` when nothing changes, and then it's complicated/inefficient to implement `foo!` with `foo`. But `scrub!` seems to never return `nil`.

**Comment by LillianZ** [comment]
Thanks for the insight! Apparently just adding `return self if valid_encoding?` suffices to fix my test suite, but I'm still digging in to see if I can turn what happened into a spec.

**Comment by eregon** [comment]
I guess one difference that can happen is without that fix `replace(dup)` would be done, which notably loses any state on `self` that's not in the String's bytes like instance variables. Potentially C extensions might also notice it if they had a `RSTRING_PTR()`, but that would be a buggy C-ext because it should not assume the `RSTRING_PTR()` does not change after calling `scrub!`.

Could you add a spec testing that `scrub!` preserves instance variables if the String is already valid?
That way we make sure we don't regress there.

**Comment by LillianZ** [comment]
Added specs for freeze, taint, and instance variables!
The spec that broke before was the one with frozen strings; the replace causes "FrozenError: can't modify frozen String"

File: spec/ruby/core/string/scrub_spec.rb [hunk 2]
```diff
@@ -105,4 +105,25 @@
     input.scrub! { |b| "<?>" }
     input.should == "a<?>"
   end
+
+  it "does not maintain the state of tainted strings that are already valid" do
```
Comment (eregon) [inline spec/ruby/core/string/scrub_spec.rb:109]: That description is confusing, I believe what happens is that `taint` does nothing on Ruby 2.7+.
I'd recommend just removing this spec, `taint` is deprecated/noop anyway in 2.7+.

Comment (LillianZ) [reply spec/ruby/core/string/scrub_spec.rb]: good idea; thanks!

**Review by eregon — APPROVED** [review]

**Pull request merged**